File: Ruby/merge_sort.rb
def merge(left_array, right_array)
    if right_array.empty?
        return left_array
    end

    if left_array.empty?
        return right_array
    end

    smallest_number = if left_array.first <= right_array.first
        left_array.shift
    else
        right_array.shift
    end

    recursive = merge(left_array, right_array)

    [smallest_number].concat(recursive)
end

def merge_sort(array)
    size = array.length
    if size <= 1
        return array
    end

    mid = (size / 2).round
    left = array.take(mid)
    right = array.drop(mid)

    sorted_left = merge_sort(left)
    sorted_right = merge_sort(right)

    return merge(sorted_left, sorted_right)
end

File: Python/Algorithm/merge_sort.py
def merge(left, right):
    merged = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1
    
    while i < len(left):
        merged.append(left[i])
        i += 1
    
    while j < len(right):
        merged.append(right[j])
        j += 1
    
    return merged


def merge_sort(array):
    if len(array) <= 1:
        return array
    
    mid = len(array) // 2

    left = merge_sort(array[:mid])
    right = merge_sort(array[mid:])

    return merge(left, right)

data = [64, 34, 25, 12, 22, 11, 90]
sorted_data = merge_sort(data)
print(sorted_data) 